### PR TITLE
Add CMakeList.txt

### PR DESCRIPTION
**Pull request by ankurvdev**

### CODE REVIEW

**Comment by ankurvdev** [comment]
Ported from https://github.com/libusb/libusb/pull/1134

File: CMakeLists.txt [hunk 1]
```diff
@@ -1,5 +1,167 @@
-cmake_minimum_required(VERSION 3.16 FATAL_ERROR)
```
Comment (Youw) [inline CMakeLists.txt:1]: - unless you have a specific motivation, I specifically suggest to limit the minimum version at 3.16
- cmake recommends to have as the very fist line of the script a `cmake_minimum_required`; it sets the default policies which might have affect on the `project` command;

File: CMakeLists.txt [hunk 2]
```diff
@@ -1,5 +1,167 @@
...
+check_include_files(sys/types.h      HAVE_SYS_TYPES_H)
+check_include_files(unistd.h         HAVE_UNISTD_H)
+check_include_files("stdlib.h;stdarg.h;string.h;float.h" STDC_HEADERS)
+
+check_symbol_exists(EFD_CLOEXEC     "sys/eventfd.h" HAVE_DECL_EFD_CLOEXEC)
+check_symbol_exists(EFD_NONBLOCK    "sys/eventfd.h" HAVE_DECL_EFD_NONBLOCK)
+check_symbol_exists(TFD_CLOEXEC     "sys/eventfd.h" HAVE_DECL_TFD_CLOEXEC)
+check_symbol_exists(timerfd_create  "sys/timerfd.h" HAVE_TIMERFD) 
+
+check_struct_has_member("struct timespec" tv_sec time.h HAVE_STRUCT_TIMESPEC)
```
Comment (Youw) [inline CMakeLists.txt:44]: since we will try to make this repo well supported, I suggest to make it right since the start
so instead of porting the Autotools config file as is - lets only include what is really used by the libusb

File: CMakeLists.txt [hunk 3]
```diff
@@ -1,5 +1,167 @@
...
+
+if(("${CMAKE_CXX_COMPILER_ID}" STREQUAL Clang) OR ("${CMAKE_CXX_COMPILER_ID}" STREQUAL GNU))
+    check_c_compiler_flag("-fvisibility=hidden" HAVE_VISIBILITY)
+endif()
+
+if(HAVE_VISIBILITY)
+    set(DEFAULT_VISIBILITY "__attribute__((visibility(\"default\")))"  CACHE INTERNAL "")
+else()
+    set(DEFAULT_VISIBILITY ""  CACHE INTERNAL "")
+endif()
```
Comment (Youw) [inline CMakeLists.txt:72]: again - I don't believe `CACHE` is required

File: CMakeLists.txt [hunk 4]
```diff
@@ -1,5 +1,167 @@
...
+        "${libusb_ROOT}/os/events_posix.c"
+        "${libusb_ROOT}/os/threads_posix.c"
+        "${libusb_ROOT}/os/linux_usbfs.c"
+        "${libusb_ROOT}/os/linux_udev.c"
+
+        "${libusb_ROOT}/os/events_posix.h"
+        "${libusb_ROOT}/os/threads_posix.h"
+        "${libusb_ROOT}/os/linux_usbfs.h"
+    )
+    target_link_libraries(usb-1.0 PUBLIC udev)
```
Comment (Youw) [inline CMakeLists.txt:139]: this has to be PRIVATE

File: CMakeLists.txt [hunk 5]
```diff
@@ -1,5 +1,167 @@
...
+        "${libusb_ROOT}/os/threads_posix.c"
+        "${libusb_ROOT}/os/linux_usbfs.c"
+        "${libusb_ROOT}/os/linux_netlink.c"
+        "${libusb_ROOT}/os/linux_android_jni.c"
+
+        "${libusb_ROOT}/os/events_posix.h"
+        "${libusb_ROOT}/os/threads_posix.h"
+        "${libusb_ROOT}/os/linux_usbfs.h"
+    )
+    target_include_directories(usb-1.0 PUBLIC android)
```
Comment (Youw) [inline CMakeLists.txt:155]: thie has to be PRIVATE

File: CMakeLists.txt [hunk 6]
```diff
@@ -1,5 +1,167 @@
...
+)
+
+if(libusb_BUILD_SHARED_LIBS)
+    add_library(usb-1.0 SHARED ${common_sources})
+else()
+    add_library(usb-1.0 STATIC ${common_sources})
+endif()
+
+target_include_directories(usb-1.0 PUBLIC "${libusb_ROOT}")
+target_include_directories(usb-1.0 PUBLIC "${CMAKE_CURRENT_BINARY_DIR}/gen_include")
```
Comment (Youw) [inline CMakeLists.txt:102]: this has to be PRIVATE

File: CMakeLists.txt [hunk 7]
```diff
@@ -1,5 +1,167 @@
...
+    "${CMAKE_CURRENT_BINARY_DIR}/gen_include/config.h"
+)
+
+if(libusb_BUILD_SHARED_LIBS)
+    add_library(usb-1.0 SHARED ${common_sources})
+else()
+    add_library(usb-1.0 STATIC ${common_sources})
+endif()
+
+target_include_directories(usb-1.0 PUBLIC "${libusb_ROOT}")
```
Comment (Youw) [inline CMakeLists.txt:101]: we need an option to make it `SYSTEM PUBLIC` or just `PUBLIC`

**Review by Youw — COMMENTED** [review]
a general comment:

- I suggest we follow some general CMake guidelines/code-style, e.g.: `UPPER_CAMEL_CASE` for all variables, including user-defined;

---
At the end, I'd like us to some practices we've umplemented for HIDAPI (documented [here](https://github.com/libusb/hidapi/blob/master/BUILD.cmake.md)), e.g. `add_subdirectory(libusb)` should be easy to do for any project (be configurable)

File: CMakeLists.txt [hunk 8]
```diff
@@ -1,5 +1,167 @@
...
+    target_compile_definitions(usb-1.0 PRIVATE HAVE_LIBUDEV=1)
+endif()
+
+if(ANDROID)
+    target_sources(usb-1.0 PRIVATE
+        "${libusb_ROOT}/os/events_posix.c"
+        "${libusb_ROOT}/os/threads_posix.c"
+        "${libusb_ROOT}/os/linux_usbfs.c"
+        "${libusb_ROOT}/os/linux_netlink.c"
+        "${libusb_ROOT}/os/linux_android_jni.c"
```
Comment (bansan85) [inline CMakeLists.txt:149]: linux_android_jni doesn't exist anymore

File: CMakeLists.txt (hunk 5 above)
Comment (bansan85) [inline CMakeLists.txt:155]: It's `target_link_libraries` and you should add `PUBLIC log`

File: CMakeLists.txt [hunk 9]
```diff
@@ -1,5 +1,167 @@
...
+# Explicitly disable use of response files
+set(CMAKE_C_USE_RESPONSE_FILE_FOR_INCLUDES   OFF)
+
+include(CheckIncludeFiles)
+include(CheckFunctionExists)
+include(CheckSymbolExists)
+include(CheckStructHasMember)
+include(CheckCCompilerFlag)
+include(CTest)
+
```
Comment (bansan85) [inline CMakeLists.txt:15]: To proper install *Config.cmake and *Target.cmake, add

```
include(GNUInstallDirs)
include(CMakePackageConfigHelpers)
```

File: CMakeLists.txt [hunk 10]
```diff
@@ -1,5 +1,167 @@
...
+if(libusb_BUILD_TESTING AND BUILD_TESTING)
+    add_executable(testlibusb ${libusb_ROOT}/../tests/testlib.c ${libusb_ROOT}/../tests/stress.c)
+    target_link_libraries(testlibusb PRIVATE usb-1.0)
+    add_test(testlibusb COMMAND testlibusb)
+endif()
 
-message(FATAL_ERROR "Finish me!")
+if(libusb_INSTALL_TARGETS)
+    install(TARGETS usb-1.0)
```
Comment (bansan85) [inline CMakeLists.txt:165]: Replace these two lines with:

```

    configure_package_config_file(
        libusbConfig.cmake.in
        ${CMAKE_CURRENT_BINARY_DIR}/libusbConfig.cmake
        INSTALL_DESTINATION ${CMAKE_INSTALL_DIR}/libusb-1.0)

    write_basic_package_version_file(
        ${CMAKE_CURRENT_BINARY_DIR}/libusbConfigVersion.cmake
        COMPATIBILITY SameMajorVersion)
    install(
        TARGETS usb-1.0
        EXPORT libusbTargets
        LIBRARY DESTINATION ${CMAKE_INSTALL_LIBDIR}
        INCLUDES
        DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}/libusb-1.0)

    install(
        EXPORT libusbTargets
        FILE libusbTargets.cmake
        NAMESPACE libusb::
        DESTINATION ${CMAKE_INSTALL_LIBDIR}/cmake/libusb-1.0)
    install(FILES ${CMAKE_CURRENT_SOURCE_DIR}/../libusb/libusb.h DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}/libusb-1.0)
    install(FILES ${CMAKE_CURRENT_BINARY_DIR}/libusbConfig.cmake
        ${CMAKE_CURRENT_BINARY_DIR}/libusbConfigVersion.cmake
        DESTINATION ${CMAKE_INSTALL_LIBDIR}/cmake/libusb-1.0)
```

And add a file `libusbConfig.cmake.in` with content:

```
@PACKAGE_INIT@

include(CMakeFindDependencyMacro)

include("${CMAKE_CURRENT_LIST_DIR}/libusbTargets.cmake")

check_required_components(libusb)
```

File: CMakeLists.txt [hunk 11]
```diff
@@ -1,5 +1,167 @@
...
+
+    "${CMAKE_CURRENT_BINARY_DIR}/gen_include/config.h"
+)
+
+if(libusb_BUILD_SHARED_LIBS)
+    add_library(usb-1.0 SHARED ${common_sources})
+else()
+    add_library(usb-1.0 STATIC ${common_sources})
+endif()
+
```
Comment (bansan85) [inline CMakeLists.txt:100]: Since I install/export in a `NAMESPACE libusb`, you should add an alias for local linking:

```
add_library(libusb::usb-1.0 ALIAS usb-1.0)
```

File: CMakeLists.txt [hunk 12]
```diff
@@ -1,5 +1,167 @@
-cmake_minimum_required(VERSION 3.16 FATAL_ERROR)
-
 project(usb-1.0 C)
```
[the remark this answers is not in the document]
Comment (Youw) [reply CMakeLists.txt:1]: even better - we should find a way to parse that from libusb sources

File: CMakeLists.txt (hunk 5 above)
Comment (Youw) [reply CMakeLists.txt:155]: `PRIVATE log` - it is not a part of public API

File: CMakeLists.txt (hunk 9 above)
Comment (Youw) [reply CMakeLists.txt:15]: I agree on this, but suggest making it as a separate PR - it is a good  chunk of work to make a propper CMake export package

File: CMakeLists.txt (hunk 10 above)
Comment (Youw) [reply CMakeLists.txt:165]: yeap, that's what I suggest to make as a separate PR
@bansan85 do I see a volountier here?

File: CMakeLists.txt (hunk 11 above)
Comment (Youw) [reply CMakeLists.txt:100]: next PR

File: CMakeLists.txt (hunk 7 above)
[the remark this answers is not in the document]
Comment (bansan85) [reply CMakeLists.txt:101]: But maybe in another MR since it's will be part of the installation feature. `INSTALL_INTERFACE` is equivalent to `PUBLIC SYSTEM include/libusb...`

File: CMakeLists.txt (hunk 5 above)
Comment (bansan85) [reply CMakeLists.txt:155]: I made it PUBLIC because I needed it in my project but I think you're right. I probably forget to add it in my project.

File: CMakeLists.txt (hunk 10 above)
Comment (bansan85) [reply CMakeLists.txt:165]: Sure. But I don't think it's possible to do a MR for an MR. I will have to wait this MR is merged ?

File: CMakeLists.txt (hunk 7 above)
[the remark this answers is not in the document]
Comment (Youw) [reply CMakeLists.txt:101]: I don't believe it is
`BUILD_INTERFACE` vs `INSTALL_INTERFACE` are independent from `PUBLIC SYSTEM`

File: CMakeLists.txt (hunk 10 above)
Comment (Youw) [reply CMakeLists.txt:165]: yes, lets wait and finish one chunk of work

File: CMakeLists.txt [hunk 13]
```diff
@@ -1,5 +1,167 @@
...
+option(libusb_INSTALL_TARGETS "Install libusb targets" ON)
+option(libusb_BUILD_TESTING "Build Tests" ON)
+option(ENABLE_LOGGING "Enable Logging" ON)
+
+if(WIN32)
+    set(PLATFORM_WINDOWS 1 CACHE INTERNAL "")
+endif()
+if(UNIX)
+    set(PLATFORM_POSIX 1  CACHE INTERNAL "")
+endif()
```
[the remark this answers is not in the document]
Comment (ankurvdev) [reply CMakeLists.txt:62]: I vaguely remember cache being required for anything that needs to used in the configure_file
I'm not a cmake expert by any means so it could be that i'm doing something else wrong

File: CMakeLists.txt [hunk 14]
```diff
@@ -1,5 +1,167 @@
...
+        "${libusb_ROOT}/os/linux_android_jni.c"
+
+        "${libusb_ROOT}/os/events_posix.h"
+        "${libusb_ROOT}/os/threads_posix.h"
+        "${libusb_ROOT}/os/linux_usbfs.h"
+    )
+    target_include_directories(usb-1.0 PUBLIC android)
+endif()
+
+if(libusb_BUILD_TESTING AND BUILD_TESTING)
```
[the remark this answers is not in the document]
Comment (ankurvdev) [reply CMakeLists.txt:158]: It is mainly for if i'm ( as a consumer of external project libusb) doing a add_subdirectory for libusb and while i want to BUILD_TESTING for my project but i dont want it for external projects like libusb to start building their tests

**Comment by ankurvdev** [comment]
There's quite a bit of comments in here :) and i'll have limited bandwidth until next week.
I do agree with most of them and it's probably good to accommodate all of them before a merge.

I'd really appreciate any contributions to the branch if anyone else has bandwidth and would like to accelerate the merge (Seems like there's a few followup that might be queued behind it)

File: CMakeLists.txt (hunk 14 above)
[the remark this answers is not in the document]
Comment (Youw) [reply CMakeLists.txt:158]: my impression that only `libusb_BUILD_TESTING` would be enough for that

**Comment by ankurvdev** [comment]
> a general comment:
> 
> - I suggest we follow some general CMake guidelines/code-style, e.g.: `UPPER_CAMEL_CASE` for all variables, including user-defined;
> 
> ---
> At the end, I'd like us to some practices we've umplemented for HIDAPI (documented [here](https://github.com/libusb/hidapi/blob/master/BUILD.cmake.md)), e.g. `add_subdirectory(libusb)` should be easy to do for any project (be configurable)


It's worth adding a cmake style guide here since I have my own preferences but I'll gladly follow what you'd like them to be


I tend to follow lower_case for local scoped variables and function names and UPPER_CASE for global scope variables / cached variables (I probably have violations for my own policy in this pr) but a few lines to standardize styling might help everyone.

We can probably add a comment at the top here for styling if you don't want an extra file

File: CMakeLists.txt (hunk 14 above)
[the remark this answers is not in the document]
Comment (Youw) [reply CMakeLists.txt:158]: > libusb devs use BUILD_TESTING

In reality, I'd open `cmake-gui` (or `ccmake`) - find there a `libusb_BUILD_TESTING` options and be very suprised why setting it to ON doesn't do anything.

In this case a uniform behavior (i.e. always use `libusb_BUILD_TESTING` only) would be more convenient (less surprising for sure).

**Comment by Youw** [comment]
> I probably have violations for my own policy

That's probably what triggered my attention the most.

> We can probably add a comment at the top here for styling if you don't want an extra file

Ideally would be to introduce something like [cmake-format](https://cmake-format.readthedocs.io/en/latest/configuration.html). But for now, jsut keeping the implementation clean/consistent - worth more as far as efforts spending is conserned.

File: CMakeLists.txt (hunk 14 above)
[the remark this answers is not in the document]
Comment (ankurvdev) [reply CMakeLists.txt:158]: Sure ok. Let's go with one variable. I never use cmake-gui so that never occurred to me

**Comment by bansan85** [comment]
@ankurvdev 
> I'd really appreciate any contributions to the branch if anyone else has bandwidth and would like to accelerate the merge (Seems like there's a few followup that might be queued behind it)

It's complicated to contribute on a PR because I can't push code on it. Should I create a PR on our repo and you merge it ?

**Comment by Youw** [comment]
I'm pretty sure you can push directly into ankurvdev/libusb-cmake:master
If not - you may ask @ankurvdev to add you directly as a contributor.

Preferably to keep the entire conversation in this thread.
_Unless_ @ankurvdev is using his master branch elsewhere, in thich case messing around with it might be inconvenient for him/her.

**Comment by ankurvdev** [comment]
Let me know if there are any permission issues. The branch should be open for any contributions.

**Comment by madwax** [comment]
Hi I might have gotten a little carried away. A Thursday night little project went south because libusbs lack of cmake support and I pulled this repo to find the "finish me" message ;) Anyway while doing other things I've updated the CMakefile per the comments.  There were also a few things wrong or missing so I've sorted them out.  @ankurvdev you should have a pull request on your branch if you want to take it forward.

**Comment by ankurvdev** [comment]
Thanks a lot @madwax. This looks a lot clean
I think you've addressed most of the comments already but I added a few touch ups.

**Comment by ankurvdev** [comment]
@Youw 
I'm wondering what your thoughts are on using subtree (https://www.atlassian.com/git/tutorials/git-subtree) for managing the repository rather than a submodule.

I don't have too much experience with subtree but I used it in a project and liked it. 
It does require a little discipline to keep things simple (for eg: just based on my limited experience)
* Forbidding local commits to upstream
* Keeping the folder hierarchy same (commit history becomes harder to navigate since paths dont line up)

But in the end the experience for users and downstream projects would be quite neat.
* Users wont have to worry about recursive submodule
* The project layout here would be essentially same as libusb with the added file (CMakeLists.txt in the root)

we might even petition vcpkg to switch to this repo for expanding support for libusb to more platforms

**Comment by mcuee** [comment]
> Thanks a lot @madwax. This looks a lot clean I think you've addressed most of the comments already but I added a few touch ups.

Nice one. The PR is now working pretty well.

One minor issue that the static lib name is not so correct for MSVC build. It should be `libusb-1.0.lib` and not `usb-1.0.lib`. MinGW build is correct -- the name of the static lib is `libusb-1.0.a`.

```
PS C:\work\libusb\libusb_cmake_pr1> cmake -B build_vs2022
-- Building for: Ninja
-- The C compiler identification is MSVC 19.35.32216.1
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.35.32215/bin/Hostx64/x64/cl.exe - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Looking for clock_gettime
-- Looking for clock_gettime - not found
-- Looking for pthread_condattr_setclock
-- Looking for pthread_condattr_setclock - not found
-- Looking for pthread_setname_np
-- Looking for pthread_setname_np - not found
-- Looking for pthread_threadid_np
-- Looking for pthread_threadid_np - not found
-- Looking for eventfd
-- Looking for eventfd - not found
-- Looking for pipe2
-- Looking for pipe2 - not found
-- Looking for syslog
-- Looking for syslog - not found
-- Looking for include file asm/types.h
-- Looking for include file asm/types.h - not found
-- Looking for include file string.h
-- Looking for include file string.h - found
-- Looking for include file sys/time.h
-- Looking for include file sys/time.h - not found
-- Looking for timerfd_create
-- Looking for timerfd_create - not found
-- Looking for nfds_t
-- Looking for nfds_t - not found
-- Performing Test HAVE_STRUCT_TIMESPEC
-- Performing Test HAVE_STRUCT_TIMESPEC - Success
-- Configuring done
-- Generating done
-- Build files have been written to: C:/work/libusb/libusb_cmake_pr1/build_vs2022
PS C:\work\libusb\libusb_cmake_pr1> cmake --build build_vs2022
[34/34] Linking C executable xusb.exe

PS C:\work\libusb\libusb_cmake_pr1> ls .\build_vs2022\


    Directory: C:\work\libusb\libusb_cmake_pr1\build_vs2022


Mode                 LastWriteTime         Length Name
----                 -------------         ------ ----
d-----          9/4/2023   6:30 pm                CMakeFiles
d-----          9/4/2023   6:30 pm                gen_include
-a----          9/4/2023   6:30 pm           4312 .ninja_deps
-a----          9/4/2023   6:30 pm           2905 .ninja_log
-a----          9/4/2023   6:30 pm          46140 build.ninja
-a----          9/4/2023   6:30 pm          14856 CMakeCache.txt
-a----          9/4/2023   6:30 pm           1930 cmake_install.cmake
-a----          9/4/2023   6:30 pm         341504 dpfp.exe
-a----          9/4/2023   6:30 pm         992312 dpfp.ilk
-a----          9/4/2023   6:30 pm        1708032 dpfp.pdb
-a----          9/4/2023   6:30 pm         356864 fxload.exe
-a----          9/4/2023   6:30 pm        1015824 fxload.ilk
-a----          9/4/2023   6:30 pm        1748992 fxload.pdb
-a----          9/4/2023   6:30 pm          28952 getoptMSVC.lib
-a----          9/4/2023   6:30 pm         327168 hotplugtest.exe
-a----          9/4/2023   6:30 pm         968696 hotplugtest.ilk
-a----          9/4/2023   6:30 pm        1683456 hotplugtest.pdb
-a----          9/4/2023   6:30 pm         326656 listdevs.exe
-a----          9/4/2023   6:30 pm         965016 listdevs.ilk
-a----          9/4/2023   6:30 pm        1683456 listdevs.pdb
-a----          9/4/2023   6:30 pm         328704 sam3u_benchmark.exe
-a----          9/4/2023   6:30 pm         972584 sam3u_benchmark.ilk
-a----          9/4/2023   6:30 pm        1683456 sam3u_benchmark.pdb
-a----          9/4/2023   6:30 pm         330752 testlib.exe
-a----          9/4/2023   6:30 pm         981912 testlib.ilk
-a----          9/4/2023   6:30 pm        1691648 testlib.pdb
-a----          9/4/2023   6:30 pm         332288 testlibusb.exe
-a----          9/4/2023   6:30 pm         966848 testlibusb.ilk
-a----          9/4/2023   6:30 pm        1683456 testlibusb.pdb
-a----          9/4/2023   6:30 pm        1052556 usb-1.0.lib
-a----          9/4/2023   6:30 pm         366592 xusb.exe
-a----          9/4/2023   6:30 pm         997736 xusb.ilk
-a----          9/4/2023   6:30 pm        1724416 xusb.pdb
```

**Comment by mcuee** [comment]
Same for shared build with `-D LIBUSB_BUILD_SHARED_LIBS=ON`. 
MinGW build is correct to produce `libusb-1.0.dll` and `libusb-1.0.dll.a`. 
MSVC build is incorrect to produce `usb-1.0.dll` and `usb-1.0.lib`.

**Comment by mcuee** [comment]
Other than the above mentioned issues, the PR already works pretty well. I tested under Windows using the following test enviroment.
1) Developer PowerShell for VS 2022, using Ninja or `Visual Studio 17 2022` generator
2) Using VS2022 to open the folder, which can create configurations for different platforms like ARM64, x86, x64, etc
3) Using MinGW64 without MSYS2, using Ninja or `MinGW Makefiles` generator
4) Using MinGW64 with MSYS2, using Ninja or `MSYS Makefiles` generator

**Comment by madwax** [comment]
@mcuee About the lib name, I would say it's in fact correct, in over +20 year of coding I've never seen a Windows static lib prefixed with lib unless there is a Linux coder around ;) The lib prefix is very much a POSIX think which is why MinGW64 does it.  However naming the bin is easy.

Also good spot on the test being stress by bad.

**Comment by madwax** [comment]
@ankurvdev Just a quick question, why did you put back the "" around all the paths? Did you run into a problem or was that just a style thing?

**Comment by ankurvdev** [comment]
Quotes around paths in general are needed to handle paths with spaces.

Most of source paths here don't need spaces anymore since I yanked out the SOURCE_DIR from LIBSUSB_ROOT which wasn't really needed and libusb paths themselves dont have spaces, but I also now simply consider it a style+good-practice to have all paths and single path variables be enclosed in Quotes for uniformity.
That way, variables that can have multiple paths (lists) also stand out.

**Comment by madwax** [comment]
Good point on the spaces and it's its a standards thing then cool.

The CMAKE_CURRENT_SORUCE_DIR Is something I do because of problems with nesting CMakefiles so its a defensive thing, in this case it might be overkill.

**Comment by ankurvdev** [comment]
I'm wondering when that's useful. 

I've never had to, the only place I remember using it was an add_custom_commands that did some sort of code generation etc and wanted to deal exclusively in absolute paths with Binary dir as my working dir.

I tried an uber project with libusb add_subdirectory, with both , it being an actual subdirectory and with absolute path outside the outer project (with binary_dir) and it worked.

So I figured I'll remove it and let you holler back :)

**Comment by madwax** [comment]
lol problem.
The issue is when you want to do any processing of a targets source files.  For example I have functions that prettify a targets folder structure in VS and XCode, when you get the source files and they are relative then SOURCE_DIR is used to build the abs path which is a bugger as that's set to current CMakefile.  I use similar functions to create pseudo-targets for static analysis, styling etc etc.

**Comment by Youw** [comment]
> I'm wondering what your thoughts are on using subtree

Haven't tried it yet, but I believe this is a perfect opportunity to do so for me. Git submodules is the default option under the tips of my fingers, but I'm in constant search of the alternatives to it.

> I would say it's in fact correct, in over +20 year of coding I've never seen a Windows static lib prefixed with lib unless there is a Linux coder around

I totally get you you on this one. In fact - I usually recommend using the "default" platform-specific behavior, which is familiar to the "natives" of the platform.

But in this particular case - we're trying to mimc the current Autotools build system of libusb, so that in cases like this there is no or minimum difference in the behavior.

TL;DR: Lets set the [`PREFIX`](https://cmake.org/cmake/help/latest/prop_tgt/PREFIX.html) property for libusb target to `lib` unconditionally for all platforms, same way as it is done in current build system(s) for libusb.

**Comment by Youw** [comment]
Thanks for all of the updates. I'll have a chance to take a look, hopefully, later this week.

**Comment by Youw** [comment]
> @Youw
I'm wondering what your thoughts are on using subtree

See my thoughts at: https://github.com/libusb/libusb-cmake/discussions/3

**Comment by ankurvdev** [comment]
Ping ... 
Should we merge this now ?

File: CMakeLists.txt [hunk 15]
```diff
@@ -1,5 +1,273 @@
...
+  DESCRIPTION "A cross-platform library to access USB devices"
+  VERSION ${LIBUSB_VERSION}
+  LANGUAGES C )
+
+if(MINGW)
+  # This is because CMake on mingw can incorrectly use response files for includes
+  # instead of passing them directly via -I<dir>
+  # Explicitly disable use of response files
+  set(CMAKE_C_USE_RESPONSE_FILE_FOR_INCLUDES OFF)
+endif()
```
Comment (Youw) [inline CMakeLists.txt:26]: my strong oppinion on this, that things like this shouldn't be a part of project CMake files
I've used CMake with MinGW in the past, and never had to use this particular option and had no issues w/o it
that's the beauty of CMake: those who need it can set it up externally, and everyone else - should care that this option even exists
```suggestion
```

File: CMakeLists.txt [hunk 16]
```diff
@@ -1,5 +1,273 @@
...
+    check_include_files(sys/eventfd.h    HAVE_EVENTFD)
+    check_include_files(string.h         HAVE_STRING_H)
+    check_include_files(sys/time.h       HAVE_SYS_TIME_H)
+
+    check_symbol_exists(timerfd_create  "sys/timerfd.h" HAVE_TIMERFD)
+    check_symbol_exists(nfds_t  "poll.h" HAVE_NFDS_T)
+
+    check_struct_has_member("struct timespec" tv_sec time.h HAVE_STRUCT_TIMESPEC)
+
+    if(HAVE_VISIBILITY)
```
Comment (Youw) [inline CMakeLists.txt:56]: I believe this flag is used before it is defined

File: CMakeLists.txt [hunk 17]
```diff
@@ -1,5 +1,273 @@
...
+        set(DEFAULT_VISIBILITY "__attribute__((visibility(\"default\")))")
+    else()
+        set(DEFAULT_VISIBILITY "" )
+    endif()
+
+    # Set vars that will be written into the config file.
+    if(CMAKE_SYSTEM_NAME STREQUAL "Windows")
```
Comment (Youw) [inline CMakeLists.txt:63]: is there a particular reason here not to use the standard `WIN32` CMake variable?
```suggestion
    if(WIN32)
```

File: CMakeLists.txt [hunk 18]
```diff
@@ -1,5 +1,273 @@
...
+        set(PLATFORM_WINDOWS 1)
+    else()
+        set(PLATFORM_POSIX 1)
+    endif()
+
+    if(LIBUSB_ENABLE_LOGGING)
```
Comment (Youw) [inline CMakeLists.txt:69]: again, the option is used before it is properly defined

File: CMakeLists.txt [hunk 19]
```diff
@@ -1,5 +1,273 @@
...
+        "${LIBUSB_ROOT}/os"
+)
+
+if (LIBUSB_TARGETS_INCLUDE_USING_SYSTEM)
+    target_include_directories(usb-1.0 SYSTEM PUBLIC "${LIBUSB_ROOT}")
+else()
+    target_include_directories(usb-1.0 PUBLIC "${LIBUSB_ROOT}")
+endif()
+
+if(CMAKE_SYSTEM_NAME STREQUAL "Windows")
```
Comment (Youw) [inline CMakeLists.txt:134]:
```suggestion
if(WIN32)
```

Comment (Youw) [reply CMakeLists.txt:134]: ok, I guess the reason is to make it conssitent in all places (e.g. there is no default variable for `BSD` as of CMake 3.16, but the problem is that `WIN32` also includes `CMAKE_SYSTEM_NAME` == `WindowsCE`, `WindowsStore`, `WindowsPhone` which definitely not a POSIX systems

File: CMakeLists.txt [hunk 20]
```diff
@@ -1,5 +1,273 @@
...
+    )
+    target_link_libraries(testlib PRIVATE usb-1.0)
+    target_include_directories(testlib PRIVATE "${CMAKE_CURRENT_BINARY_DIR}/gen_include")
+    target_compile_definitions(testlib PRIVATE $<$<C_COMPILER_ID:MSVC>:_CRT_SECURE_NO_WARNINGS=1>)
+    add_test(testlib COMMAND testlib)
+endif()
+
+if(LIBUSB_BUILD_EXAMPLES)
+
+    if(CMAKE_C_COMPILER_ID STREQUAL "MSVC")
```
Comment (Youw) [inline CMakeLists.txt:216]:
```suggestion
    if(MSVC)
```

File: CMakeLists.txt [hunk 21]
```diff
@@ -1,5 +1,273 @@
...
+        add_library(getoptMSVC STATIC
+            "${LIBUSB_ROOT}/../msvc/getopt/getopt.h"
+            "${LIBUSB_ROOT}/../msvc/getopt/getopt.c"
+            "${LIBUSB_ROOT}/../msvc/getopt/getopt1.c"
```
Comment (Youw) [inline CMakeLists.txt:220]: if we have to use `../`, maybe `LIBUSB_ROOT` is chose poorly in a first place

File: config.h.in [hunk 22]
```diff
@@ -0,0 +1,75 @@
...
+#cmakedefine USE_SYSTEM_LOGGING_FACILITY
+
+/* Enable GNU extensions. */
+#cmakedefine _GNU_SOURCE 1
+
+/* Define to `__inline__' or `__inline' if that's what the C compiler
+   calls it, or to nothing if 'inline' is not supported under any name.  */
+#ifndef __cplusplus
+/* #undef inline */
+#endif
```
Comment (Youw) [inline config.h.in:75]: is this some unfinished development?

File: CMakeLists.txt (hunk 16 above)
Comment (Youw) [reply CMakeLists.txt:56]: ah, OK, i see - it is inside a function called after the variable is defined...

File: CMakeLists.txt (hunk 21 above)
Comment (Youw) [reply CMakeLists.txt]: but lets keep as is for now

File: config.h.in (hunk 22 above)
Comment (Youw) [reply config.h.in:75]: #4

File: CMakeLists.txt (hunk 2 above)
Comment (Youw) [reply CMakeLists.txt]: #5

File: CMakeLists.txt [hunk 23]
```diff
@@ -1,5 +1,269 @@
-cmake_minimum_required(VERSION 3.16 FATAL_ERROR)
+cmake_minimum_required(VERSION 3.16)
 
-project(usb-1.0 C)
+# define the source root
```
Comment (Youw) [inline CMakeLists.txt:3]:
```suggestion
```

File: CMakeLists.txt [hunk 24]
```diff
@@ -1,5 +1,268 @@
...
+        "${LIBUSB_ROOT}/os/windows_usbdk.c"
+        "${LIBUSB_ROOT}/os/windows_usbdk.h"
+        "${LIBUSB_ROOT}/os/windows_winusb.c"
+        "${LIBUSB_ROOT}/os/windows_winusb.h"
+        $<$<C_COMPILER_ID:MSVC>:${LIBUSB_ROOT}/libusb-1.0.rc>
+    )
+    target_compile_definitions(usb-1.0 PRIVATE $<$<C_COMPILER_ID:MSVC>:_CRT_SECURE_NO_WARNINGS=1>)
+    target_link_libraries(usb-1.0 PRIVATE windowsapp)
+else()
+    # common POSIX/non-Windows system
```
Comment (Youw) [inline CMakeLists.txt:153]:
```suggestion
    # common POSIX/non-Windows sources
```

**Review by Youw — APPROVED** [review]

**Comment by Youw** [comment]
I've applied all of the fixes from my comments myself, and I think we're in a good place to have it merged.

**Comment by Youw** [comment]
> Same for shared build with `-D LIBUSB_BUILD_SHARED_LIBS=ON`. MinGW build is correct to produce `libusb-1.0.dll` and `libusb-1.0.dll.a`. MSVC build is incorrect to produce `usb-1.0.dll` and `usb-1.0.lib`.

Fixed.
